src/hooks/useForm.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useState, useMemo } from 'react';
+
+// Interfaces START
+
+interface Validator {
+  (value: string): string;
+}
+
+interface FormInput {
+  field: string;
+  value: string;
+  validators: Validator[];
+}
+interface ParameterFunction<T> {
+  (value: FormInput): T;
+}
+
+interface Dictionary<T> {
+  [key: string]: T;
+}
+
+// Interfaces END
+
+// Hook START
+const useForm = (settings: FormInput[]) => {
+  const createObjectFrom = (
+    arraySource: FormInput[],
+    keyFunction: ParameterFunction<string>,
+    valueFunction: ParameterFunction<any>
+  ) =>
+    arraySource.reduce((object: Dictionary<any>, item: FormInput) => {
+      object[keyFunction(item)] = valueFunction(item);
+      return object;
+    }, {});
+
+  const calculateErrorList = (
+    value: string,
+    errorFunctions: Validator[]
+  ): string[] => {
+    let errorArray: string[] = [];
+    for (let errorFunction of errorFunctions) {
+      let errorMessage: string = errorFunction(value);
+      if (errorMessage) {
+        errorArray.push(errorMessage);
+      }
+    }
+    return errorArray;
+  };
+
+  const errorFunctionsByKey: Dictionary<[]> = useMemo(() => {
+    return createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.validators || []
+    );
+  }, []);
+
+  const [fields, setFields] = useState<Dictionary<string>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => item.value || ''
+    )
+  );
+  const [errors, setErrors] = useState<Dictionary<[]>>(
+    createObjectFrom(
+      settings,
+      (item) => item.field,
+      (item) => calculateErrorList(item.value, errorFunctionsByKey[item.field])
+    )
+  );
+
+  const disableForm: boolean = useMemo(() => {
+    return !![].concat(...Object.values(errors)).length;
+  }, [errors]);
+
+  const onChange = (event: React.ChangeEvent<{ name: any; value: any }>) => {
+    const { name, value } = event.target;
+
+    if (errorFunctionsByKey[name].length) {
+      const newErrorList: string[] = calculateErrorList(
```

Comment:
As you might know, TS has a technique to infer certain types when dealing with functions returned values stored in variables.
Take a look at this example:
`
function sum(a: number, b: number): number {
  return a * b;
}
const total = sum(10, 2);
`
Whenever we called a function that has a specific return type, in this case sum() : number  and stored that value returned in a variable, that variable is very likely to have the type inferred (correctly) by typescript. In this case, when hovering over total, type number is assigned.

Going back to the commented code: calculateErrorList function has a return type defined, so when you do 
`const newErrorList: string[] = calculateErrorList`, typescript will aready know that newErrorList is an array of string, because that is the type you defined for that function.

The conclusion: it is a good practice to always let inferred the types whenever TS can. 

* Example where type annotation is well received :) *
NOTE: Sometimes TS can not inferred the type correctly because the function has different behaviours depending on the arguments passed. Take a look at the JSON.parse(arg) function. It can return maany value types, depending on the arg passed
`
// JSON.parse('false') // Output: boolean
// JSON.parse('4') // Output: number
// JSON.parse('{ name: 'stefano '}') // Output: { name: string }
// JSON.parse('{ age: 'stefano '}') // Output: { age: number }
`

This is a good case to EXPLICITLY set the type to a variable that will stored the value return by JSON.parse(arg)
`const coordinatesAnnotation: { x: number; y: number } = JSON.parse(
    "{ x: number, y: number }"
);`

Reply:
Ready, thanks 👯